### PR TITLE
feat: add sodium-native

### PR DESCRIPTION
**Pull request by tuyennhv**
**Motivation**
- Add sodium-native as a new crypto implementation

**Description**
- Started from https://github.com/web3-storage/ipfs-elastic-provider-bitswap-peer/blob/main/src/noise-crypto.js, thanks @ShogunPanda 
- Add sodium-native as a new crypto implementation
- Benchmark shows that stablelib is still way faster than (~6x) sodium-native probably due to the cost to call native layer

```
stablelib decrypt x 268,550 ops/sec ±1.05% (90 runs sampled)
sodium-native decrypt x 45,880 ops/sec ±2.47% (88 runs sampled)
stablelib encrypt x 259,846 ops/sec ±0.70% (92 runs sampled)
sodium-native encrypt x 39,432 ops/sec ±3.09% (82 runs sampled)
```

### CODE REVIEW

**Comment by tuyennhv** [comment]
with a `NOISE_MSG_MAX_LENGTH_BYTES` of 65535 I got the following result

| # |Buffer size|stablelib (opts/sec)|sodium-native (opts/sec)|
|---|----------|---------|--------------|
|decrypt|100|321,615|26,026|
|encrypt|100|199,691|41,907|
|decrypt|500|211,464|24,126|
|encrypt|500|93,179|38,781|
|decrypt|5000|43,205|18,972|
|encrypt|5000|13,326|29,192|
|decrypt|20000|12,318|13,803|
|encrypt|20000|3,404|15,123|
|decrypt|65535|3,809|6,729|
|encrypt|65535|1,044|5,552|

for decryption, sodium-native's performance flips at a Buffer size of 20000
for encryption, sodium-native's performance flips at a Buffer size of around 3000-4000

File: benchmarks/crypto.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  const data = Buffer.from('encryptthis encryptthis encryptthis encryptthis')
+  const nonce = 1000
+  const nonceBytes = new Uint8Array(12)
+  new DataView(nonceBytes.buffer, nonceBytes.byteOffset, nonceBytes.byteLength).setUint32(4, nonce, true)
+  const key =  new Uint8Array(Array.from({length: 32}, () => 1))
+  const encrypted = stablelib.chaCha20Poly1305Encrypt(data, nonceBytes, new Uint8Array(0), key)
+
+  for(const {id, crypto} of [
+    {id: 'stablelib decrypt', crypto: stablelib},
+    {id: 'sodium-native decrypt', crypto: sodiumNative}
```
[the remark this answers is not in the document]
Comment (tuyennhv) [reply benchmarks/crypto.js]: @dapplion for eth2, most of the chunk size is 0-500, per 100 chunks like that there's a big chunk size at 20000. I updated the benchmark below

**Comment by ShogunPanda** [comment]
Please note that the crypto operations might not be the only meaningful metric.
Stablelib blocks the event loop as it is JS only, while sodium does not and thus the overall server concurrency might increase even for small buffers.

**Comment by tuyennhv** [comment]
> Please note that the crypto operations might not be the only meaningful metric. Stablelib blocks the event loop as it is JS only, while sodium does not and thus the overall server concurrency might increase even for small buffers.

@ShogunPanda my understanding is sodium also blocks the event loop unless it provides some async apis

**Comment by tuyennhv** [comment]
This profile result shows that `sodium-native` has a lot of javascript-native communication overhead
[sodium-native-prof-result.txt](https://github.com/ChainSafe/js-libp2p-noise/files/8965801/sodium-native-prof-result.txt)

**Comment by tuyennhv** [comment]
decided to go with https://github.com/ChainSafe/as-chacha20poly1305

**Pull request closed without merging**